### PR TITLE
Add mixin version when building the bundle

### LINKED ISSUE #1948: Record the mixin version used to build the bundle
Right now we are including a list of the names of the mixins used to build a bundle, but not their version.

**current bundle.json**
```json
{
  "custom": {
    "sh.porter": {
            "commit": "db1b9092",
            "manifestDigest": "cbdb1783bbf6aa55d0ac4a4e5a1245f3bd43a811556259b2034539292464501d",
            "mixins": {
                "exec": {},
                "helm3": {},
                "kubernetes": {}
            }
    }
  }
}
```

**desired bundle.json**
```json
{
  "custom": {
    "sh.porter": {
            "commit": "db1b9092",
            "manifestDigest": "cbdb1783bbf6aa55d0ac4a4e5a1245f3bd43a811556259b2034539292464501d",
            "mixins": {
                "exec": { "version": "v1.0.0-alpha.9" },
                "helm3": { "version": "v0.1.15-5-g25cda36" },
                "kubernetes": { "version": "v0.12.4" }
            }
    }
  }
}
```

This change needs to be made against the release/v1 branch in the following files: 
https://github.com/getporter/porter/blob/ecedc49b353f12deaf2573c48e45dbc663583ee6/pkg/cnab/config-adapter/stamp.go#L77-L81

https://github.com/getporter/porter/blob/ecedc49b353f12deaf2573c48e45dbc663583ee6/pkg/porter/build.go#L152-L169

### PR DESCRIPTION
**Pull request by joshuabezaleel**
# What does this change
Append the version of the mixin when building the bundle.

# What issue does it fix
Closes #1948 

# Notes for the reviewer

# Checklist
- [ ] Did you write tests?
- [ ] Did you write documentation?
- [ ] Did you change porter.yaml or a storage document record? Update the corresponding schema file.
- [ ] If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

# Reviewer Checklist
* Comment with /azp run test-porter-release if a magefile or build script was modified
* Comment with /azp run porter-integration if it's a non-trivial PR

[contributors]: https://porter.sh/src/CONTRIBUTORS.md

### CODE REVIEW

**Comment by joshuabezaleel** [comment]
Uploading 1948-bundle-mixin-version.mp4…

@carolynvs There are still some things that I think I need confirmation for:
1. I am still not sure what to change on the `pkg/porter/build.go` since the change on `pkg/cnab/config-adapter/stamp.go` already achieved the goal of adding the mixin version to the `bundle.json` file as the result of the bundle's build process.
2. I am not yet writing any test since I am not sure what version to put on the expected of the assertion statement since the version of the mixin will differ for each of the local machine. 

Looking forward to the review, Carolyn. Thank you lots in advance!

**Comment by carolynvs** [comment]
Oh I'm so glad that you picked this up, thanks! 👍

**Review by carolynvs — COMMENTED** [review]
This looks right, maybe for the tests try running `mage UpdateTestFiles` and see if they start passing?

**Comment by joshuabezaleel** [comment]
I ran `mage UpdateTestFiles` but the test still fails. 
There's one suspicion that I have yet to confirm and find a way to resolve.

I tried to print the `ManifestConverter` and its generated `stamp` and try to compare its different output between the one at `TestConfig_GenerateStamp` on `pkg/cnab/config-adapter/stamp_test.go` and to one running `porter build` command using the [`whalesay` bundle](https://github.com/getporter/porter/blob/main/examples/docker/porter.yaml). 
image
image

The `Mixins` part both on `ManifestConverter` and `stamp` was empty on the `TestConfig_GenerateStamp` test compared to the `porter build` output which I suspect was caused by the mixins' metadata on invoking [`NewManifestConverter` ](https://github.com/getporter/porter/blob/release/v1/pkg/cnab/config-adapter/stamp_test.go#L26) was supplied by `nil` (the 4th input parameter), compared to the [one ](https://github.com/getporter/porter/blob/release/v1/pkg/porter/build.go#L184)at the build process which was supplied by the output of `getUsedMixins()`.

1. [`getUsedMixins()`](https://github.com/getporter/porter/blob/release/v1/pkg/porter/build.go#L156) was `Porter`'s struct method so I feel conflicted if I need to instantiate a `NewTestPorter` at `pkg/cnab/config-adapter/stamp_test.go` since it should only test things related to package `pkg/cnab`.
2. I just remembered that you mentioned on the [issue](https://github.com/getporter/porter/issues/1948) that the `getUsedMixins` function should be changed as well, but the bundles' mixins' version is already on the `bundle.json` just by the change on the `/pkg/cnab/config-adapter/stamp.go` as was shown on the video I've attached earlier. What should be changed on this `getUsedMixins`, if I may ask? 

So sorry that it took me a really long time to reply to this one, Carolyn. Looking forward to hearing from you.

File: pkg/cnab/config-adapter/stamp.go [hunk 1]
```diff
@@ -75,9 +78,11 @@ func (c *ManifestConverter) GenerateStamp() (Stamp, error) {
 	stamp.EncodedManifest = base64.StdEncoding.EncodeToString(rawManifest)
 
 	// Remember the mixins used in the bundle
-	stamp.Mixins = make(map[string]MixinRecord, len(c.Manifest.Mixins))
-	for _, m := range c.Manifest.Mixins {
-		stamp.Mixins[m.Name] = MixinRecord{}
+	stamp.Mixins = make(map[string]MixinRecord, len(c.Mixins))
```
Comment (carolynvs) [inline pkg/cnab/config-adapter/stamp.go:81]: Here you are looping over the list of installed mixins, not the list of mixins used in the bundle. You need to loop over `c.Manifest.Mixins` which are the mixins used by the bundle. Then match the mixin name against the list of installed mixins `c.Mixins` to lookup the version number.

Feel free to rename c.Mixins to c.InstalledMixins to help clear up confusion between those two variables.

The other problem is that the existing unit tests that test this functionality are not passing in a list of installed mixins. You'll need to change the tests like so:

```go
     installedMixins := []mixin.Metadata{
		{Name: "exec", VersionInfo: pkgmgmt.VersionInfo{Version: "v1.2.3"}},
	}
	a := NewManifestConverter(c.Context, m, nil, installedMixins)
```

Comment (carolynvs) [reply pkg/cnab/config-adapter/stamp.go:81]: Once you are done, you either need to update the test files with `mage updateTestFiles` or edit pkg/cnab/config-adapter/testdata/mybuns.bundle.json by hand so that it includes the mixin version number.

**Review by carolynvs — CHANGES_REQUESTED** [review]
I left some comments with a few more changes we need to make before this is ready

**Comment by carolynvs** [comment]
@joshuabezaleel Just a general request for pull requests going forward, when possible do not use rebase after someone has started reviewing it.

When you add new commits that incorporate the requested feedback, then I can review just the new commit you pushed. When rebase is used, the reviewer has to review the entire PR from scratch since every commit has changed. Rereading every single file to make sure that they are correct, every time you make a change makes the review process a lot work more for the reviewers.

Don't worry about lots of commits, you can either let me know that you want to clean up the commits once the review is complete, or what I usually do is squash the PR when I merge it so that it's a single commit.

**Comment by joshuabezaleel** [comment]
@carolynvs Ah, truly sorry for the inconvenience. I really thought at the beginning it was the other way around, that rebasing would make it more helpful for the reviewer, since it keeps the log straight.

> When rebase is used, the reviewer has to review the entire PR from scratch since every commit has changed. Rereading every single file to make sure that they are correct, every time you make a change makes the review process a lot work more for the reviewers.

But yeah, I haven't thought of this earlier. Makes perfect sense. 
Permission to confirm, so if there's conflict, I can do as follows, right? 
```
git checkout release/v1 && git pull upstream release/v1
git checkout branch && git merge release/v1
# resolve conflict
git add . && git commit -m "Pull latest changes and resolve conflict" && git push origin branch
```

Now that I learned it gave the opposite effect, will definitely keep this in mind for future PRs. Thank you lots for the feedback, Carolyn!

File: pkg/cnab/config-adapter/stamp.go [hunk 2]
```diff
@@ -109,7 +118,7 @@ func (c *ManifestConverter) DigestManifest() (string, error) {
 	v := pkg.Version
 	data = append(data, v...)
 
-	for _, m := range c.Mixins {
+	for _, m := range c.InstalledMixins {
```
[the remark this answers is not in the document]
Comment (joshuabezaleel) [reply pkg/cnab/config-adapter/stamp.go]: Updated 😄

**Comment by carolynvs** [comment]
Yup, doing a merge and resolving the conflicts that way are totally fine.

File: pkg/cnab/config-adapter/stamp.go [hunk 3]
```diff
@@ -109,8 +114,9 @@ func (c *ManifestConverter) DigestManifest() (string, error) {
 	v := pkg.Version
 	data = append(data, v...)
 
-	for _, m := range c.Mixins {
-		data = append(append(data, m.Name...), m.Version...)
+	usedMixinsVersion := c.getUsedMixinsVersion()
+	for usedMixinName, usedMixinVersion := range usedMixinsVersion {
+		data = append(append(data, usedMixinName...), usedMixinVersion...)
```
Comment (carolynvs) [inline pkg/cnab/config-adapter/stamp.go:119]: I think the way we are concatenating these bytes is already pretty odd but I don't want to have you try to fix that in this PR since it's unrelated. Once your PR is in, I'll submit a PR to clean that up.

**Review by carolynvs — APPROVED** [review]
:shipit: Thanks for sticking with it and getting this fix in!

File: pkg/cnab/config-adapter/stamp.go (hunk 3 above)
Comment (carolynvs) [reply pkg/cnab/config-adapter/stamp.go:119]: It's just hard to read and when thinking about the digest as a whole, hard to follow what we are hashing here.
